File: remove-nth-node-from-end-of-list.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode RemoveNthFromEnd(ListNode head, int n) {
        int size = 0;
        for(ListNode c = head; c != null; c = c.next)
            size++;
        
        if(size == 1 || (size - n) == 0)
            return head.next;

        ListNode prev = head, curr = head.next;
        for(int i = 1, target = size - n; i < target; i++){
            prev = curr;
            curr = curr.next;
        }
        
        prev.next = curr.next;

        return head; 
    }
}

File: remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *endFinder, *curr, *prev;
        for(endFinder = head; n--; endFinder = endFinder->next);
        
        for(prev = curr = head; endFinder; endFinder = endFinder->next)
            prev = curr, curr = curr->next;

        if(prev == curr)
            return head->next;
        
        prev->next = curr->next;

        return head;
    }
};